File: scala/src/prob004/Prob004.scala
package prob004

/**
 * A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.
 * Find the largest palindrome made from the product of two 3-digit numbers.
 */
object Prob004 extends App {

  def isPalindrom(i: Int): Boolean = i.toString.reverse.toInt == i

  val res = (for {
    x <- 10 to 1000
    y <- 10 to 1000
    if isPalindrom(x*y)
  } yield (x, y, x*y)).maxBy(_._3)

  println(res)
  // -> (913,993,906609)
}

File: java/src/prob004/Prob004.java
package prob004;

import util.Palindrom;

public class Prob004 {

    /**
     * A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.
     * Find the largest palindrome made from the product of two 3-digit numbers.
     */
    public static void main(String... args) {

        int max = 0;
        for (int i=100; i<1000; i++) {
            for (int j=100; j<1000; j++) {
                int k = i*j;
                if (k > max && Palindrom.isPalindrom(k)) {
                    System.out.println(i + " * " + j + " = " +k);
                    max = k;
                }
            }
        }

        // 913 * 993 = 906609
    }
}
